### ISSUE TITLE
See if we can use Python properties to get/set DOT attributes

### ISSUE THREAD

**Issue by peternowee** [pydot/pydot#250]
This is a reminder issue that during development of pydot 2.0.0 we should see if we can use [Python `property`](https://docs.python.org/3.9/library/functions.html#property) for getters/setters of DOT attributes. This was suggested by @johnyf in https://github.com/pydot/pydot/issues/151#issuecomment-353866822 and by @sandrotosi in https://github.com/pydot/pydot/issues/217#issuecomment-536327992 for example.

Some points of attention already:

- [ ] Currently (pydot 1.4.2.dev0), we dynamically create getter/setter methods, also known as convenience methods, like `Node.get_color()`, `Edge.set_label()`, etc. These seem to be used a lot, so we will need to have a generous transition/deprecation period for those.
- [ ] Currently (pydot 1.4.2.dev0), we have no clear way to prevent name clashes between pydot attributes and (custom) DOT attributes. For example, `Node.set_name()` and `Graph.set_type()` set pydot attributes, so getter/setter methods created for custom DOT attributes `name` and `type` would clash with those. Perhaps when introducing new getters/setters based on Python properties, we can somehow provide separate namespaces for them, at least for the (custom) DOT attributes over which we have no control. For example by prefixing getters/setters for (custom) DOT attributes with `dot_`: `node.dot_color = 'blue'` (DOT attribute used by Graphviz), `node.dot_name = 'My Name'`, `graph.dot_type = 'My Type'` (custom DOT attributes). The main drawback of this of course the inconvenience of having to use the `dot_`-prefix all the time, so other suggestions are welcome.
- [ ] There are over 200 DOT attributes for which we need to generate getters/setters. Even when using Python properties, it may be desirable to dynamically generate them instead of having a long source file with duplicate code to set up `property` for each DOT attribute individually. In PR pydot/pydot#242, also targeted for pydot 2.0.0, I proposed changes to how the current getter/setter methods are dynamically created. We will have to see if the dynamic generation of Python properties can fit in there as well.

**Comment by ferdnyc** [comment]
> * Perhaps when introducing new getters/setters based on Python properties, we can somehow provide separate namespaces for them, at least for the (custom) DOT attributes over which we have no control. For example by prefixing getters/setters for (custom) DOT attributes with `dot_`: `node.dot_color = 'blue'` (DOT attribute used by Graphviz), `node.dot_name = 'My Name'`, `graph.dot_type = 'My Type'` (custom DOT attributes). The main drawback of this of course the inconvenience of having to use the `dot_`-prefix all the time, so other suggestions are welcome.

My suggestion, on this, would be to simply put all of the attributes inside a member object of the class. So, e.g.:

```python
node = pydot.Node()
node.attrib.color = 'blue'
node.attrib.style = 'myclass'
```

At face value it doesn't seem much different than the `dot_` prefixes, but there are several advantages:

1. Dotted names are easier to type than underscore prefixes.

1. Someone who's really averse to the extra typing can easily alias the attributes member object:
   ```python
   node = pydot.Node()
   a = node.attrib
   a.color = 'blue'
   a.style = 'myclass'
   ```

1. Similarly, it's far easier to script changes to member attributes than prefixed names.
   ```python
   node = pydot.Node()
   attributes = {'color': 'blue', 'style': 'myclass'}
   for k, v in attributes.items():
       setattr(node.attrib, k, v)
   ```

**Comment by ferdnyc** [comment]
Oh, and as an added bonus, if `node.attrib` implements the [Mapping interface](https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping) (basically, has implementations for the abstract `__len__`, `__iter__`, and `__getitem__` methods, and is a subclass of `collections.abc.Mapping`), then it can also be used as a dictionary:

```python3
node = pydot.Node()
node.attrib["color"] = "blue"
attributes_dict = dict(node.attrib)
attribute_names = node.attrib.keys()
```

**Comment by ajnelson-nist** [comment]
Hello!  I'm chiming in as a random user[^1].  FWIW, I think it would be beneficial for static type-review to know when a Dot property I'm trying to set or get is being set on the right class of GraphViz thing.  I often forget if a property on a node is also available on a graph, or if there's a dialect restriction like on [`len`](https://graphviz.org/docs/attrs/len/).

I think the suggestion for `dot_` prefixes on the Python getters and setters is reasonable disambiguation.  Though, out of curiosity, how many clashes between `pydot` properties and GraphViz properties are there?

[^1]: Disclaimer: Participation by NIST in the creation of the documentation of mentioned software is not intended to imply a recommendation or endorsement by the National Institute of Standards and Technology, nor is it intended to imply that any specific software is necessarily the best available for the purpose.

**Comment by lkk7** [comment]
@ajnelson-nist NIST people are definitely not random. They are special!

Static, untyped autocomplete might be doable, because there's already a list of attributes (possibly outdated) in the code. This would probably require code generation, or a lot of manual work. After that initial work is done, we'd need to keep the declarations in sync with the "ground truth" list. It would also add a lot of lines to the project, but it could be all hidden in a separate file. Mixing that with @ferdnyc's suggestion above results in something like:
```
  class NodeAttributes(AttributeMapping):
      color: Any
      shape: Any
      label: Any
      ...
```

But to make that typed correctly and include dialect restrictions, we need to go into https://github.com/pydot/pydot/issues/352 and that's even more effort. I don't think it's realistic for me right now. As discussed in that issue, even though graphviz is rather stable, you never know which version or dialect someone is using.

In short, the maintenance burden would be high. It would be easier if pydot targeted a specific Graphviz release, but it doesn't.

As for collisions, I think there are no collisions between [current official attributes](https://graphviz.org/doc/info/attrs.html) and pydot's `set_...` functions.

**Comment by ajnelson-nist** [comment]
Thank you for the thought and kind words.

I definitely respect the maintenance burden.  Drift monitoring isn't necessarily easy.

Re:

> [...] even though graphviz is rather stable, you never know which version or dialect someone is using.

Naïvely (modulo Liskov Substitution Principle), I'd think subclassing would help manage the dialect angle, e.g.:

```python
class DotGraph(Graph):
    ...
    def add_edge(graph_edge: Edge, *args: Any, **kwargs: Any):
        if not isinstance(graph_edge, DotEdge):
            raise TypeError("...")
        super().add_edge(*args, **kwargs)
class NeatoGraph(Graph):
    def add_node(graph_node: Node) -> None:
        if not isinstance(graph_node, NeatoNode):
            raise TypeError("...")
        super().add_node(*args, **kwargs)
```

It looks like the LSP-induced `isinstance` checking would actually be consistent with the runtime type checking being done in `Graph.add_edge` and `.add_node`.

Meanwhile, while I'm personally a fan of `enum`s like mentioned in 352, I also see the trade in strength of (type) typing vs. strength of (hands) typing, both for `pydot` maintainers and `pydot` users.

In any case, thank you again for the reply.

**Comment by ferdnyc** [comment]
Part of the problem here, also, is that there is no actual fixed list of attributes for _any_ part of Graphviz.

There's a fixed list of attributes that **each layout engine** will recognize and interpret, but in theory the user _should_ be free to add ANY random, free-form attributes they want. There is nothing wrong with a DOT file like this:

```gv
graph G {
  rankdir = LR;
  bgcolor = "#BBFFBB80";
  copyright = "Graph Examples, Ltd.";
  copyright_year = 2026;
  a -> b [penwidth=5, shape=square, designer="Example Guy"];
}
```

...Making that work with fixed attribute lists gets tricky, because you kind of still want to accept a value set on **any** key, even one that's not known to the code.

**Comment by ferdnyc** [comment]
> Naïvely (modulo Liskov Substitution Principle), I'd think subclassing would help manage the dialect angle

Hm. Subclassing to bake-in the selection of layout engine is an interesting idea.

It's sort of "anti-graphviz" in a way, because part of the idea with the dotfile format is that you _can_ (at least in theory; the option is _there_ even though it's probably made use of pretty infrequently) run the same graph definition through different layout engines, saving that choice until the end. Some engines may **ignore** some attributes, but that's OK because having them in the file doesn't hurt anything when using the others. Ignoring something is not the same as rejecting or prohibiting it. (Back to that whole, "there are no practical restrictions on what attributes can be set on any graph element" thing again.)

I'd think the interface would have to be pretty nice to make the whole subclass deal not-annoying.

Personally, I already find it annoying that you have to create `Node` or `Edge` objects just to add elements to graphs, especially since **we don't store them that way**. The first thing we do is throw away that object and just keep its dictionary. So, why force the user to create the object?

Really, in my ideal world, since _this_ works:
```python
import pydot
g = pydot.Dot("G", graph_type="graph")
n = pydot.Node("A", shape="square", penwidth=2)
g.add_node(n)
```

then, _this_ should work as well:
```python
import pydot
g = pydot.Dot("G", graph_type="graph")
g.add_node("A", shape="square", penwidth=2)
```

Cut out the middleman.

Those two listings should have identical effect; there's no reason `add_node()` can't create a `Node` object for you. Or, if people don't like overloading the method, then we could at least offer a `.create_node()` that takes `Node()` constructor arguments instead. (And so on for `.create_edge()`, `.create_subgraph()`, etc.)

Now, if we had _that_ interface, then the subclassing wouldn't be so bad, because the `.add/create_node()` of a `NeatoGraph` would know to create `NeatoNode`s, etc. (I addressed the need for graph objects to be able to create properly-typed elements, specifically to enable subclassing, in #495 which isn't merged yet but is basically done.)

But having to create `NeatoNode`s and `NeatoEdge`s, specifically, to be able to add them to a `NeatoGraph` feels... well, I don't know, maybe it's not really any more clunky than my already-clunky first code listing above, which of course is also our current, actual interface.

**Comment by ajnelson-nist** [comment]
@ferdnyc : What would you expect to happen in this sequence of calls, adapting your example?

```python
import pydot
g = pydot.Dot("G", graph_type="graph")
g.add_node("A", shape="square")
na = pydot.Node("A", shape="circle")
g.add_node(na)
```

What would node `A`'s shape be?  Or should this raise an error because of an ID collision?

**Comment by ferdnyc** [comment]
@ajnelson-nist Same thing that would happen with this:

```python
import pydot
g = pydot.Dot("G", graph_type="graph")
n = pydot.Node("A", shape="square")
g.add_node(n)
na = pydot.Node("A", shape="circle")
g.add_node(na)
```

...it's not like how you construct the Node objects changes anything about how they're handled.

And what would happen today, with either version, is this:
```python
>>> print(g.to_string(indent=2))
graph G {
  A [shape="square"];
  A [shape="circle"];
}
```

...Which `dot` would render as a graph containing a single, circular `A` node. Later node specifications override previous ones for the same node.

**Edit:** I should say, subsequent node specs _modify_ previous ones, not override. It's accumulative. This will be a circular blue node, for example:

```dot
graph G {
  A [shape=square, color="blue"];
  A [shape=circle];
}
```

**Comment by ajnelson-nist** [comment]
@ferdnyc : Re: "There's a fixed list of attributes that **each layout engine** will recognize and interpret, [...]":

This sounds to me like a question of parsing strictness, specializable to each engine.  The attributes _not_ recognized by whichever engine could be stashed in a dictionary, free to be `dict[str, Any]` or firmer; the attributes that _are_ are recognized could enjoy firmer type & content review.

**Comment by ajnelson-nist** [comment]
@ferdnyc :

> ... Later node specifications [modify] previous ones for the same node.

Thanks, I'd forgotten the conflict resolution strategy at GraphViz-render time.  I see how that kind of sets a parallel behavior expectation in Python space.

**Comment by ferdnyc** [comment]
@ajnelson-nist Yeah, the thing to be aware of regarding pydot as it exists _today_, is that it's not really a graph library. It would be more accurate to describe it as a dotfile syntax generator. See #493 for an extremely rambling tour through all the implications of _that_.

But it also affects the attribute/typing issue. Because pydot just generates _dotfile syntax_, it really only deals in string values, because its only **real** output is the result of `pydot.Graph.to_string()`. The fact that the `K` attribute to `Graph` and `Cluster` is interpreted — **only** by `fdp` and `sfdp` — as a double (float) numeric value in the range `0.0 - +∞`... that's all defined by Graphviz.

**IF** you're willing to lock in the choice of layout engine up front, then it becomes _feasible_ to bake in enforcement of those restrictions, but it risks making the API kind of weird.

Like, say `FdpGraph` had a type-annotated `self.attrib.K: float | None`. But `DotGraph` and `NeatoGraph` wouldn't, so would you end up having to do this?

```python
import pydot
g = pydot.FdpGraph("G")
g.attrib.K = 5.2
h = pydot.DotGraph("H")
h.extra_attribs["K"] = 5.2
```

The asymmetry feels offputting, to me.

And what about the parser (_our_ parser)? Today it doesn't really parse to any specific rendering engine, and couldn't unless we also define _parsing syntax_ for each one, **and** force the user to choose before parsing any data. So how does `pydot.dot_parser.graphparser` interpret this syntax?

```python
graph G {
  K = 5.2;
}
```
That's  **always** a valid graph definition, but how it's interpreted by pydot suddenly becomes dependent on which layout engine you plan to run it through, which would have to be specified prior to even parsing the data.

**Comment by ajnelson-nist** [comment]
I think the asymmetry's a matter of preference.  I personally like knowing there's only one place I should look for a certain property, regardless of the path the data takes to get there.  On a general `Graph` class, I'm inclined to agree there might not be a great answer except maybe making a getter/setter pair for every known attribute of every dialect, but then subclassing for dialects would lead to...maybe `NotImplementedError`s?  I forget if that's just for TODOs.

```python
class Node:
    # Aside - the "dot_" prefix suggested above in the thread
    # would probably need to generalize to "gv_" for the sake
    # of other dialects.
    @property
    def gv_K(self) -> float | None: ...

class DotNode(Node):
    @property
    def gv_K(self) -> float | None:
        raise NotImplementedError("K not available in Dot nodes.")
```

As for parsing: Maybe a Rust idiom could be borrowed here.  If there's a serializer `pydot.Graph.to_string()`, what about having a deserializer `@classmethod`, `pydot.Graph.from_string() -> Graph`?  Subclassing seems to work without raising any LSP issues.  I just ran the below through a type checker.

```python
# This software was developed at the National Institute of Standards
# and Technology by employees of the Federal Government in the course
# of their official duties. Pursuant to Title 17 Section 105 of the
# United States Code, this software is not subject to copyright
# protection within the United States. NIST assumes no responsibility
# whatsoever for its use by other parties, and makes no guarantees,
# expressed or implied, about its quality, reliability, or any other
# characteristic.
#
# We would appreciate acknowledgement if the software is used.


class ClassA:
    @classmethod
    def from_string(cls, s: str) -> "ClassA":
        return ClassA()


class ClassB(ClassA):
    @classmethod
    def from_string(cls, s: str) -> "ClassB":
        return ClassB()


def foo(x: ClassA) -> None:
    print("It worked!")


def main() -> None:
    obj_1 = ClassB.from_string("")
    foo(obj_1)


if __name__ == "__main__":
    main()
```

I just offer that idea for the sake of how high-level API could look.  I have no comments on the internals down to `pydot.dot_parser.graphparser`.
